VFB-197: Add audit log to delete user action

minalui opened the pull request:
## What's changed
- Add audit log when user is deleted

## Screenshots / Videos
| Before     | After      |
|------------|------------|
| paste_here | paste_here |

## Checklist
- [x] The ticket is up-to-date - Please document any deviations from the original approach if there is any.
- [x] I have documented the testing steps for QA
- [x] I have self-reviewed this PR
- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've verified it works via `npm run build` and `npm run start`
- [x] Make sure you've fixed all linting problems with `npm run lint_fix`
- [x] Make sure you've tested via `npm run test:e2e`

If you have made any changes to the database... N/A
  - [x] The migration files are up-to-date with my final set up (`npx supabase db diff -f <name_of_migration>` should create nothing at this point)
  - [x] I have updated the typescript definitions for the database with `db:local:generate_types`
  - [ ] I have modified the seed in `supabase/seed/seed.mts` if appropriate
  - [ ] If I have modified the seed, I have also generated the seed with `npm run db:generate_seed` 
  - [x] With my final set up, I can run `npm run dev:reset_supabase` without any errors.
  - Does this require resetting the deployed database? - YES / NO (remove as appropriate)
  - [x] I have checked that migration can happen successfully without resetting the database, doing the following.
    - Make sure you have rebased your branch onto dev or merged dev into your branch.
    - Checkout the dev branch
    - Run `npm run post_checkout` to reset the database, including the seed data.
    - Checkout your branch
    - Run `npx supabase migrations up --include-all --local` to run all outstanding migration files
    - Check that the resulting database is what you expect, bar any seed data changes.

rickaw commented on src/app/admin/deleteUser/DeleteUserDialog.tsx:
```diff
@@ -36,29 +38,59 @@ const DeleteUserDialog: React.FC<Props> = (props) => {
         }
 
-        const { error } = await adminDeleteUser(props.userToDelete.id);
+        const { data: profileUserId, error: profileUserIdError } = await supabase
```
I wonder if we should get the profile id and send the audit log within `adminDeleteUser` instead?

minalui commented:
I've added the subscription to the profiles table so the user table updates automatically. I've also removed the refresh button. However, the create user form remains populated after the submit button (and the success message is displayed instead of a submit button) - just wondering if the form should be cleared and submit button remain?

rickaw commented:
> I've added the subscription to the profiles table so the user table updates automatically. I've also removed the refresh button. However, the create user form remains populated after the submit button (and the success message is displayed instead of a submit button) - just wondering if the form should be cleared and submit button remain?
> 
> ![image](https://private-user-images.githubusercontent.com/160640328/322803307-8b3877c0-e177-42d2-b701-bd470a785517.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyOTI1NjYsIm5iZiI6MTcxMzI5MjI2NiwicGF0aCI6Ii8xNjA2NDAzMjgvMzIyODAzMzA3LThiMzg3N2MwLWUxNzctNDJkMi1iNzAxLWJkNDcwYTc4NTUxNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxNlQxODMxMDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NGMxOTc4YjUxNjBjMTA2ZTQyODg1MTkzZmExZjA1NTk1ZGYwNjBhNGVjNDgzMzlhZDlmYWNlNmFiODk0ZDkxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Xyz_IdKG5gf8ol3X9CU-q_N1teK8QgP5phC9qL3Trpo)

I think it's fine as is for now!

rickaw commented on src/app/admin/deleteUser/DeleteUserDialog.tsx:
```diff
@@ -36,9 +36,14 @@ const DeleteUserDialog: React.FC<Props> = (props) => {
+        if (deleteUserError) {
+            props.setAlertOptions({
+                success: false,
+                message: <>{`${deleteUserError.type}. Log ID: ${deleteUserError.logId}`}</>,
```
since `deleteUserError.type` is meant to be just an identifier of the error, we should compute the user friendly error message based on that and display it there :)

rickaw commented on src/app/admin/deleteUser/DeleteUserDialog.tsx:
```diff
@@ -48,17 +53,8 @@ const DeleteUserDialog: React.FC<Props> = (props) => {
             });
             void logInfoReturnLogId(`${props.userToDelete?.email} deleted successfully.`);
-        } else {
-            const logId = await logErrorReturnLogId(
-                `Error with delete: User ${props.userToDelete.email}`,
-                { error }
-            );
```
We should still log an error

minalui replied:
I've sent a log within adminDeleteUser and returned the logID to be displayed

rickaw commented on src/app/admin/usersTable/UsersTable.tsx:
```diff
@@ -230,6 +231,26 @@ const UsersTable: React.FC = () => {
 
+    useEffect(() => {
+        void fetchAndDisplayUserData();
+        // This requires that the profiles tables have Realtime turned on
```
I don't think this comment is needed anymore as we manage realtime through migrations

rickaw commented on src/app/admin/usersTable/UsersTable.tsx:
```diff
@@ -230,6 +231,26 @@ const UsersTable: React.FC = () => {
     }, [fetchAndDisplayUserData]);
 
+    useEffect(() => {
+        void fetchAndDisplayUserData();
```
why do we need to fetch here? I thought this useEffect should just run once at the beginning to try and subscribe to the db changes?

rickaw commented on src/app/admin/usersTable/UsersTable.tsx:
```diff
@@ -230,6 +231,26 @@ const UsersTable: React.FC = () => {
+        return () => {
+            supabase.removeChannel(subscriptionChannel);
+        };
+    }, [startIndex, endIndex, primaryFilters, sortState]);
```
> I thought this useEffect should just run once at the beginning to try and subscribe to the db changes?
```suggestion
    }, [fetchAndDisplayUserData]);
```

rickaw commented on src/server/adminInviteUser.ts:
```diff
@@ -88,11 +88,11 @@ export async function adminInviteUser(
             first_name: userDetails.firstName,
             last_name: userDetails.lastName,
             telephone_number: userDetails.telephoneNumber,
+            user_id: newUserData.user.id,
```
Ooh, was this another user_id related bug?

minalui replied:
Yes - it wasn't making the connection to users table and not displaying the data in the users table

rickaw replied:
oh cool - let's merge this asap so we can include it in today's release

rickaw commented on src/server/adminDeleteUser.ts:
```diff
@@ -2,30 +2,78 @@
-    error: Record<string, string> | null;
-};
+type DeleteUserErrorType =
+    | "Failed to authenticate as admin"
+    | "Failed to get user id from profiles"
+    | "Failed to delete user";
```
These tend to be camelCase or all lowercase, since they serve as a way for the code to differentiate what kind of error it is returning, not a user friendly error message. (This function shouldn't be concerned about what error message is displayed, but about returning the error data back to the caller)

e.g. `"failedToAuthenticateAsAdmin"`

rickaw commented on src/server/adminDeleteUser.ts:
```diff
@@ -2,30 +2,78 @@
+    | "Failed to delete user";
 
-export async function adminDeleteUser(userId: string): Promise<DeleteUserErrorType> {
+type DeleteUserError =
```
nitpick: this whole object is a result, which may or may not have an error inside
```suggestion
type DeleteUserResult =
```

rickaw reviewed (commented):
Nice! a few comments

minalui commented:
Can you please double check the switch because it required to return a default value.
